pkg/login/social/generic_oauth.go

```diff
@@ -412,33 +434,82 @@ func (s *SocialGenericOAuth) FetchPrivateEmail(client *http.Client) (string, err
 	return email, nil
 }
 
-func (s *SocialGenericOAuth) FetchTeamMemberships(client *http.Client) ([]int, bool) {
+// FetchTeamMemberships will attempt to get the user's team string IDs from the JWT claims using the 'team_id_attribute_path' setting.
+// If the attribute path does not find anything in the JWT claims, then the same attribute path is used to attempt to retrieve the team IDs from the OIDC '/userinfo' endpoint.
+// If the attribute path still does not find anything in the '/userinfo' response, then the same attribute path is used to attempt to retrieve the team IDs from the 'teams_url'
+// If team IDs are found, then true is returned. Otherwise, false is returned.
+func (s *SocialGenericOAuth) FetchTeamMemberships(client *http.Client) ([]string, error) {
+	var err error
+	var ids []string
+
+	if s.teamsUrl == "" {
+		ids, err = s.fetchTeamMembershipsFromDeprecatedTeamsUrl(client)
+	} else {
+		ids, err = s.fetchTeamMembershipsFromTeamsUrl(client)
+	}
+
+	if err == nil {
+		s.log.Debug("Received team memberships", "ids", ids)
+	}
+
+	return ids, err
+}
+
+func (s *SocialGenericOAuth) fetchTeamMembershipsFromDeprecatedTeamsUrl(client *http.Client) ([]string, error) {
+	s.log.Warn("[Deprecated] Switch to the use of the teams_url and team_id_attribute_path configuration options for filtering on team membership")
+
+	var response httpGetResponse
+	var err error
+	var ids []string
+
 	type Record struct {
 		Id int `json:"id"`
 	}
 
-	response, err := s.httpGet(client, fmt.Sprintf(s.apiUrl+"/teams"))
+	response, err = s.httpGet(client, fmt.Sprintf(s.apiUrl+"/teams"))
 	if err != nil {
 		s.log.Error("Error getting team memberships", "url", s.apiUrl+"/teams", "error", err)
-		return nil, false
+		return nil, err
 	}
 
 	var records []Record
 
 	err = json.Unmarshal(response.Body, &records)
 	if err != nil {
 		s.log.Error("Error decoding team memberships response", "raw_json", string(response.Body), "error", err)
-		return nil, false
+		return nil, err
 	}
 
-	var ids = make([]int, len(records))
+	ids = make([]string, len(records))
 	for i, record := range records {
-		ids[i] = record.Id
+		ids[i] = strconv.Itoa(record.Id)
+	}
+
+	return ids, nil
+}
+
+func (s *SocialGenericOAuth) fetchTeamMembershipsFromTeamsUrl(client *http.Client) ([]string, error) {
+	if s.teamIdAttributePath == "" {
+		return []string{}, nil
 	}
 
-	s.log.Debug("Received team memberships", "ids", ids)
+	var response httpGetResponse
+	var err error
+	var ids []string
+
+	response, err = s.httpGet(client, fmt.Sprintf(s.teamsUrl))
+	if err != nil {
+		s.log.Error("Error getting team memberships", "url", s.teamsUrl, "error", err)
+		return nil, err
+	}
+
+	ids, err = s.searchJSONForStringArrayAttr(s.teamIdAttributePath, response.Body)
+
+	if err != nil {
+		return nil, nil
+	}
 
-	return ids, true
+	return ids, nil
```

Comment:
Could this just be:

```go
return s.searchJSONForStringArrayAttr(s.teamIdAttributePath, response.Body)
```
?

It's a bit of a smell to receive an error and return `nil, nil`. If we need to do something specific if `s.searchJSONForStringArrayAttr` returns an error, we can create a new error in this file and use 


```go
// at the top of the file...
var AttributeError = errors.New("failed to find json attribute")

fmt.Errorf("%s: %w", err.Error(), AttributeError)

// Check for attribute error
if err := fetchTeamMembershipsFromTeamsUrl(...); err != nil {
    if errors.Is(err, AttributeError) {
        // error is an AttributeError
    }
}
```